### PR TITLE
implement schema and analyzer support for on schemas and on table

### PR DESCRIPTION
**Pull request by m-kharbat**

### CODE REVIEW

File: blackbox/docs/sql/administration/privileges.txt [hunk 1]
```diff
@@ -90,7 +101,6 @@ other tables of the ``doc`` schema.
 
 Behavior of ``GRANT``, ``DENY`` and ``REVOKE``
 ==============================================
-
```
Comment (seut) [inline blackbox/docs/sql/administration/privileges.txt]: I think this newline is valid

File: blackbox/docs/sql/administration/privileges.txt [hunk 2]
```diff
@@ -190,8 +211,14 @@ use the :doc:`/sql/reference/revoke` SQL statement, for example the
 The privileges that were granted and denied to user wolfgang on doc.books
 can be revoked, like this::
 
-    cr> REVOKE ALL ON TABLE doc.books TO wolfgang;
-    REVOKE OK, 3 row affected (... sec)
+    cr> REVOKE ALL ON TABLE doc.books FROM wolfgang;
+    REVOKE OK, 3 rows affected (... sec)
+
+``REVOKE ALL`` or ``REVOKE ALL PRIVILEGES`` will revoke all privileges
+ on the cluster from a user, it can be used like this::
```
Comment (seut) [inline blackbox/docs/sql/administration/privileges.txt]: hm don't get this. you mean revoke without an ON ?

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 3]
```diff
@@ -1383,6 +1415,18 @@ private static boolean isDistinct(SqlBaseParser.SetQuantContext setQuantifier) {
         throw new IllegalArgumentException("Unsupported ordering: " + token.getText());
     }
 
+    private static String getclazz(Token token) {
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: pls use CamelCase

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 4]
```diff
@@ -132,37 +133,68 @@ public Node visitDropUser(SqlBaseParser.DropUserContext context) {
...
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
+                return new DenyPrivilege(usernames, getclazz(context.clazz().getStart()), tableOrSchemaNames);
+            }
         } else {
             List<String> privilegeTypes = identsToStrings(context.privilegeTypes().ident());
-            return new DenyPrivilege(usernames, privilegeTypes);
+            if (context.ON() == null) {
+                return new DenyPrivilege(usernames, privilegeTypes, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: would be great to factor this out into a method e.g. `getClassAndIdentsForPrivileges(...)` which can then be used in all 3 privilege node methods

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 5]
```diff
@@ -132,37 +133,68 @@ public Node visitDropUser(SqlBaseParser.DropUserContext context) {
...
+                return new RevokePrivilege(usernames, "CLUSTER", Collections.EMPTY_LIST);
+            } else {
+                List<QualifiedName> tableOrSchemaNames = context.qname().stream().map(q -> getQualifiedName(q)).collect(toList());
+                return new RevokePrivilege(usernames, getclazz(context.clazz().getStart()), tableOrSchemaNames);
+            }
         } else {
             List<String> privilegeTypes = identsToStrings(context.privilegeTypes().ident());
-            return new RevokePrivilege(usernames, privilegeTypes);
+            if (context.ON() == null) {
+                return new RevokePrivilege(usernames, privilegeTypes, "CLUSTER", Collections.EMPTY_LIST);
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: think it makes sense to create a different constructor leaving clazz and idents out...

File: sql-parser/src/main/java/io/crate/sql/tree/PrivilegeStatement.java [hunk 6]
```diff
@@ -56,6 +62,14 @@ public boolean all() {
         return all;
     }
 
+    public List<QualifiedName> tableOrSchemaNames() {
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/tree/PrivilegeStatement.java]: what about privilegeIdents?

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 7]
```diff
@@ -66,14 +113,14 @@ private static void ensureUserManagementEnabled(@Nullable User user) {
         }
     }
 
-    private static Collection<Privilege.Type> getPrivilegeTypes(boolean all, List<String> typeNames) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: pls rebase again and don't remove optimized code but adjust your code instead ;)

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 8]
```diff
@@ -24,40 +24,87 @@
...
+    PrivilegesAnalyzedStatement analyzeDeny(DenyPrivilege node, @Nullable User user) {
         ensureUserManagementEnabled(user);
+        Collection<Privilege.Type> privilegeTypes;
+        if (node.all()) {
+            privilegeTypes = Arrays.asList(Privilege.Type.values());
+        } else {
+            privilegeTypes = parsePrivilegeTypes(node.privileges());
+        }
+
+        if (Privilege.Clazz.SCHEMA.equals(Privilege.Clazz.valueOf(node.clazz()))) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: can we move that to a generic `validatePrivilegeIdents(Clazz, idents)` code?

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 9]
```diff
@@ -97,17 +144,35 @@ private static void ensureUserManagementEnabled(@Nullable User user) {
...
+                    Privilege.Clazz.CLUSTER,
+                    null,
+                    grantor.name()
+                );
+
+                privileges.add(privilege);
+            }
+        } else {
+            for (Privilege.Type privilegeType : privilegeTypes) {
+                for (QualifiedName name : tableOrSchemaNames) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: think it makes sense to do a if (idents == null) { create 1 for the cluster } else { 1 for each ident }

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 10]
```diff
@@ -134,6 +143,12 @@ private SchemaInfo getSchemaInfo(TableIdent ident) {
         return schemaInfo;
     }
 
+    public void validate(String schemaName) {
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: think is better to call this `validateSchemaNames`

File: sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java [hunk 11]
```diff
@@ -135,22 +227,51 @@ public void testGrantWithoutUserManagementEnabledThrowsException() {
...
-        e = SQLExecutor.builder(clusterService).build();
-        e.analyze("REVOKE DQL FROM test");
+    public void testGrantOnUnknownSchemaThrowsException() {
+        expectedException.expect(SchemaUnknownException.class);
+        expectedException.expectMessage("Schema 'hoichi' unknown");
+        analyzePrivilegesStatement("GRANT DQL on schema hoichi TO user1");
+    }
+
+    @Test
+    public void testGrantOnUnknownTabkeThrowsException() {
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java]: typo Tabke

File: sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java [hunk 12]
```diff
@@ -135,22 +227,51 @@ public void testGrantWithoutUserManagementEnabledThrowsException() {
...
+        expectedException.expect(TableUnknownException.class);
+        expectedException.expectMessage("Table 'doc.hoichi' unknown");
+        analyzePrivilegesStatement("GRANT DQL on table doc.hoichi TO user1");
+    }
+
+    @Test
+    public void testRevokeFromUnknownSchemaThrowsException() {
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java]: do we have to test this for every Privilege.State ?

File: sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java [hunk 13]
```diff
@@ -135,22 +227,51 @@ public void testGrantWithoutUserManagementEnabledThrowsException() {
...
+
+    @Test
+    public void testRevokeFromUnknownTableThrowsException() {
+        expectedException.expect(TableUnknownException.class);
+        expectedException.expectMessage("Table 'doc.hoichi' unknown");
+        analyzePrivilegesStatement("Revoke DQL on table doc.hoichi FROM user1");
     }
 
     private Privilege privilegeOf(Privilege.State state, Privilege.Type type) {
         return new Privilege(state,
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/PrivilegesDCLAnalyzerTest.java]: pls call the new privilegeOf(Privilege.State state, Privilege.Type type, Privilege.Clazz clazz, String ident) method instead

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java (hunk 9 above)
Comment (m-kharbat) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: i'm not sure how this is different .. I already check for the class and create the privilege when ident is null

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java (hunk 4 above)
Comment (m-kharbat) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: this will require a new class that contains the privilege ident and class ..

Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: yes or a tuple, but eliminates duplicated code

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzedStatement.java [hunk 14]
```diff
@@ -23,6 +23,7 @@
 package io.crate.analyze;
 
 import io.crate.analyze.user.Privilege;
+import io.crate.sql.tree.QualifiedName;
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzedStatement.java]: unused import? if so pls revert the other change in that file as well

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 15]
```diff
@@ -108,6 +108,15 @@ public Schemas(Settings settings,
         return (T) info;
     }
 
+    public void validateTableName(TableIdent ident) {
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: there is already a getTableInfo() inside this class doing same stuff...

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 16]
```diff
@@ -42,22 +45,37 @@
...
+    }
+
+    PrivilegesAnalyzedStatement analyzeGrant(GrantPrivilege node, @Nullable User user) {
         ensureUserManagementEnabled(user);
+        validatePrivilegeIdents(Privilege.Clazz.valueOf(node.clazz()), node.privilegeIdents());
+
         return new PrivilegesAnalyzedStatement(node.userNames(),
-            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.GRANT));
+            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.GRANT, node.privilegeIdents(),
+                Privilege.Clazz.valueOf(node.clazz().toUpperCase(Locale.ENGLISH))));
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: afaik the clazz will always be the defined static strings which are always uppercase

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 17]
```diff
@@ -42,22 +45,37 @@
...
 
-    static PrivilegesAnalyzedStatement analyzeRevoke(RevokePrivilege node, @Nullable User user) {
+    PrivilegesAnalyzedStatement analyzeRevoke(RevokePrivilege node, @Nullable User user) {
         ensureUserManagementEnabled(user);
+        validatePrivilegeIdents(Privilege.Clazz.valueOf(node.clazz()), node.privilegeIdents());
+
         return new PrivilegesAnalyzedStatement(node.userNames(),
-            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.REVOKE));
+            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.REVOKE, node.privilegeIdents(),
+                Privilege.Clazz.valueOf(node.clazz().toUpperCase(Locale.ENGLISH))));
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: same here, no need to uppercase afaik

File: sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java [hunk 18]
```diff
@@ -42,22 +45,37 @@
...
 
-    static PrivilegesAnalyzedStatement analyzeDeny(DenyPrivilege node, @Nullable User user) {
+    PrivilegesAnalyzedStatement analyzeDeny(DenyPrivilege node, @Nullable User user) {
         ensureUserManagementEnabled(user);
+        validatePrivilegeIdents(Privilege.Clazz.valueOf(node.clazz()), node.privilegeIdents());
+
         return new PrivilegesAnalyzedStatement(node.userNames(),
-            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.DENY));
+            privilegeTypesToPrivileges(getPrivilegeTypes(node.all(), node.privileges()), user, State.DENY, node.privilegeIdents(),
+                Privilege.Clazz.valueOf(node.clazz().toUpperCase(Locale.ENGLISH))));
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/PrivilegesAnalyzer.java]: and here

File: sql/src/main/java/io/crate/metadata/Schemas.java (hunk 15 above)
Comment (m-kharbat) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: getTableInfo() calls blockedRaiseException which will throw an exception that is not handled by the ExceptionAuthorizedValidator and that the user might not be allowed to see

**Comment by m-kharbat** [comment]
@seut i'll do a follow up commit to refactor  `getTableInfo` functions since the user is not needed in both

**Pull request merged**